freqtrade/exchange/exchange.py

```diff
@@ -669,7 +675,7 @@ def _get_stake_amount_considering_leverage(self, stake_amount: float, leverage:
     def create_dry_run_order(self, pair: str, ordertype: str, side: str, amount: float,
                              rate: float, leverage: float, params: Dict = {}) -> Dict[str, Any]:
         order_id = f'dry_run_{side}_{datetime.now().timestamp()}'
-        _amount = self.amount_to_precision(pair, amount)
+        _amount = self._contract_size_to_amount(pair, self.amount_to_precision(pair, amount))
```

Comment:
In my opinion / understanding, this call is wrong.

`amount_to_precision` should be the last to be called. 
We've had multiple instances in the past where we were ordering `4.4555667` of "coin whatever" - and the exchange told us "precision is 4" - so you need to round it to `4.4555` (or `4.4556`) for this order to be placed.

This is not the case with every exchange - but to me, it means that precision should have heigher weight than "contract size" - and should therefore be called last.

Obviously the "precision is 4" can come in multiple different forms - ccxt knows 3 i think - one is "decimal places" - the other is "0.0001" (and i forgot the 3rd as it's only used by 1 or 2 exchanges) - both of which will mean the same thing - but will have a different "precision_mode" attached to the exchange class.
Details for this are in amount_to_precision - but i think the same rules do apply to futures / margin trading as well.

Reply:
The idea with this was that it's a dry run order, so I'm just removing any references to contract sizes and using the actual amounts

Reply:
just to understand this right (from a calculation point of view)
your idea was to: 
* convert the amount from 100 to 6.66
* place that order (in live)
* convert 6.66 back to 100 for dry-run?
* leave it at 6.66 for live

I'm honestly not sure if this will all play out (not this PR, but looking at it from a bigger picture), and i think we'll need a proper concept for this (which currently, i don't really see) before going further with this PR.

Please consider that
* `trade.amount` is used in basically all profit calculations
* `trade.amount` is updated with `filled`, `amount` from the order we're getting back (live, obviously)

So if i place an order with `6.66` - i expect to get 6.66 back in the order's filled column.
Updating a dry-order "back" to an amount of 100 is therefore wrong in my understanding - and the trade object should hold 6.66 (we're holding 6.66 contracts).
Dry-run should return (roughly) the same values - which would not be the case with this.

---

The bigger problem than dry-run is the following though:

While your amount is now 6.66 (assigned by `amount_to_precision`) - all calculations now are wrong (i was ordering 100 XXX at 2$ each, so i'd expect my entry cost to be 200$).
Instead, my trade object now has an amount of 6.66 - which means my (calculated) cost is now 6.66 * 2 - which is around 13$ ... quite a long spot from where we wanted to be.

Reply:
We only need to change the value that we send to the exchange, we will still analyze it as 100 of currency X, but we need to tell the exchange that we want 6.66 contracts (If a contract is worth 15.01~ of currency X). The exchange will also return amounts in how many contracts, we should convert these into the size of the underlying asset before we deal with them. The prices are all compared to the underlying asset, if the price is 62000, then it's 62000 for one of currency X, regardless of what the contract size is

Reply:
i'm not sure if the "conversion back" is the right approach - i fear that this will lead to quite some bugs, as we'll need to have this conversion in place for wherever we use an order 
This also includes "userland", as the order object is directly passed to the timeout handlers.

an alternative might be to update the order we receive back to the "right amount" (basically patching the ccxt exchange response), which is probably simpler, but potentially also slightly riskier (we need to be 100% sure which fields this applies and which doesn't).

---

I understand this is limited to some exchanges .. do you know which exchanges are _not_ impacted by this? 
if we're left with enough alternatives, we should consider ignoring this for the moment, and excluding support for these exchanges - as this pr is a far way from "all that's needed" - as it's only handling the "place order" side at the moment - but not the "way back".

Reply:
It's not needed for linear contracts on Binance, so we don't need it immediately, but for those other exchanges, it will be pretty important, so if anyone wants to use one of those exchanges, they should take a look at this PR

Reply:
so to summarize, without this, we're ONLY left with binance? 
then we'll need to find a solution to this i think. i was hoping that this would only impact 2-3 exchanges, not "most" ...

Reply:
It's not needed for FTX either, but I won't be implementing cross margin anytime soon, which is needed for FTX

Reply:
I'm not happy with a "only binance" approach - as we can lock ourselfs in a corner with this (similar to what happened to ccxt, looking at it in hindsight) - but i see this PR as high risk change, as it can mess up all calculations in very subtle and odd ways, which will need both VERY good test, as well as very good and detailed live (and dry) tests.

Then let's focus on margin trading only for the moment.
I don't see what's missing for margin trading (which will implicitly enable both leverage and short trading) - i'm sure it's not 100% complete, but it's not much either, while for futures, the list of "that's still missing" continues to become longer and longer, instead of shorter and shorter.
As such, i don't see a reason to delay finishing up margin and making that available - once that is stable, we can continue to try and implement futures.

I hinted on this earlier as well - as i think we're trying to do too much, and with that unnecessarily delaying other parts - and i regret (and excuse) for not being more decisive on this, as we could have margin available already (probably since months) if we kept our focus on that part first (which we will do now).

Reply:
To return to this discussion ... 
If we assume freqtrade will only deal with amounts - not with "contract Sizes" at any other place than when we're sending the order to the exchange - then we'll also have to convert the exchange responses back (for all values that we use as amounts - like `'filled'`, `'amount'`, (and potentially others ...?) - in both `fetch_order`, `create_order`, as well as (probably?) in `fetch_my_trades`.